### ISSUE TITLE
fused back pass

### ISSUE THREAD

**Issue by betterftr** [LoganBooker/prodigy-plus-schedule-free#1]
Hi, can you implement fused back pass (step_parameter) too so larger models can be trained too? (https://github.com/konstmish/prodigy/pull/28)

**Comment by LoganBooker** [comment]
Hey @betterftr, I can certainly take a look! Right now I'm stilling playing around with the core optimisation logic -- once that's a bit more stable I'm keen to expand the scope of memory/performance optimisations. For instance, the `foreach` version seems to performance worse than the `for` one (at least in my testing), which I'd very much like to resolve first if possible.

**Comment by betterftr** [comment]
That's amazing thank you very much, I cant wait to do tests :P

**Comment by LoganBooker** [comment]
@betterftr Alrighty, I ended up just hacking my local version of Kohya to use OneTrainer's `step_parameter` structure. Experimental support is in with this commit: https://github.com/LoganBooker/prodigy-plus-schedule-free/commit/d764afc2fd6177e5c1f87b0efa79ce28f2f4c7e7

As I suspected, the training dynamics do differ as `d` is calculated at a different point, but this didn't seem to change the end result in my limited testing.

**Comment by betterftr** [comment]
Will check tomorrow, I'll probably need to edit a lot of OT files too

**Comment by LoganBooker** [comment]
@betterftr Thanks for giving it a go. I have a very limited testing setup, so any results from other models is greatly appreciated. I just pushed a fix that makes the fused and non-fused paths match in terms of dynamics.

In terms of settings, can you try using `beta4=0` and `factored=False`? The `factored` setting is currently problematic with SD3.5 (https://github.com/konstmish/prodigy/pull/25#issuecomment-2454112866).

**Comment by betterftr** [comment]
Tried with your latest commit (Factored=False, and Beta4=0)
The lr still rises to 1.0 and stays there (orange):
![image](https://github.com/user-attachments/assets/4f5e847c-e775-4488-957e-cf125ea71da1)

**Comment by LoganBooker** [comment]
I think I know what might be happening here. Not at my machine at the moment, but will test somethings when I can. Again, thanks for your help so far!

**Comment by LoganBooker** [comment]
@betterftr Thanks for your patience. I decided to download SD3.5L myself to run my own tests rather than have you burn time on this. :P

While I do that, I've pushed another commit to bring the optimiser more in line with the reference Prodigy. If you'd like to have another go, I'd suggest using that commit and setting `split_groups=False` as well which should be mostly identical to the original Prodigy. If that doesn't work, I'll be spending my time looking a debug prints.

**Comment by betterftr** [comment]
Okay I will give it a go, with split_groups=True it wont work?
Also in summary this commit has factored, parameter slicing, schedule free and fused back pass too?

**Comment by LoganBooker** [comment]
@betterftr Yep -- I just changed how one or two calculations are done to try and track down the root cause of the issue. I'd still recommend leaving `factored=False` for now -- hopefully by training SD3.5L myself I can figure out why that's not working with other models.

**Comment by betterftr** [comment]
I'm afraid without factored true it doesnt fit on my 4090 :( well it does but its like 90s / it, can test if its working but thats not ideal training speed for quality and suchs tests

**Comment by LoganBooker** [comment]
@betterftr OK, fixed that error and added `debug_print` as an optimiser argument. Set it to `true` and it'll print which steps are being called. If you see `[Prodigy+ScheduleFree] Called step_parameter (fused backward pass).` only, that's good, but if you see `[Prodigy+ScheduleFree] Called step (regular optimiser step).` only (or both), then it means OneTrainer isn't recognising the `step_parameter`.

**Comment by betterftr** [comment]
Prodigy+ScheduleFree] Called step_parameter (fused backward pass).0:21<04:50,  2.46s/it, loss=0.158, smooth loss=0.235]
[Prodigy+ScheduleFree] Called step (regular optimiser step).

**Comment by LoganBooker** [comment]
OK, it looks like OneTrainer has hardcoded configs for the optimisers that it checks to see if they support fused backward pass: https://github.com/Nerogar/OneTrainer/blob/b5699f51c070c15c9aac340d5967fd6fa8f855f5/modules/trainer/GenericTrainer.py#L535C1-L535C113

You may need to edit this file in your setup to add Prodigy (if you're hijacking it for this optimiser): https://github.com/Nerogar/OneTrainer/blob/master/modules/util/enum/Optimizer.py#L83

It also seems to check a config object here:
https://github.com/Nerogar/OneTrainer/blob/master/modules/util/optimizer_util.py

So adding the `step_parameter` is not enough on its own, which is unfortunate -- if I'm understanding the code correctly.

**Comment by betterftr** [comment]
I edited OT's files before (see my former screenshot of optimizers settings in the gui, those weren't in originally) the only thing I missed is fused_back_pass=True in your optimizer's  __init__ I believe

**Comment by LoganBooker** [comment]
@betterftr Nope, that's completely fair. Sorry, just going off what I could see in the OT code! Anyway, I've added `fused_back_pass` to th init -- so give that a whirl.

**Comment by betterftr** [comment]
Prodigy+ScheduleFree] Called step_parameter (fused backward pass).0:16<05:03,  2.53s/it, loss=0.179, smooth loss=0.236]

okay fused seems to be working now, also not getting those absurd lr numbers, but it seems the lr stays at d0 (1e-06) now. Should  I try with beta4=0 as well? my settings are below
![image](https://github.com/user-attachments/assets/567edf46-3b22-4913-9b07-8700a081d863)

**Comment by LoganBooker** [comment]
@betterftr Yep, I think I'll need to investigate `factored` myself and test with SD3.5L to see what's going on. I have to step away now, but I've pushed a commit to only apply factorisation to 2D tensors. I don't think it will help, but worth a try. Again, thanks for your help!

**Comment by betterftr** [comment]
> Again, thanks for your help!
Its fine, I like to help. Speaking of, in dxqbYD's version setting update_clip to None / setting eps2 to 1e-25 fixed this stagnant lr in Cascade/ SD3

**Comment by betterftr** [comment]
Yes it seems to be working now! gj! I will start the quality tests, if I find anything funky can I report it back here?

**Comment by LoganBooker** [comment]
> if I find anything funky can I report it back here?
Please do!

> you should report it to the original prodigy's authors
I believe the issues have already been identified for the most part -- bfloat16 needing a slightly higher eps is understandable. I also only implement the factorised second moment from Adafactor, not the entire implementation, so they're not directly applicable.

Regarding the batch size -- can you increase `d0` to `1e-5`? I couldn't get Prodigy to kick-in using the default 1e-6 for SD3.5 (which appears to be common: https://github.com/konstmish/prodigy/issues/27). See below -- cyan is BS 1, 1e-05. Orange is BS 4, 1e-05. Purple is BS 4, 1e-06.

![image](https://github.com/user-attachments/assets/ff4ac842-55e4-49c6-b480-1181ab7c02e0)

**Comment by betterftr** [comment]
no luck :(

![image](https://github.com/user-attachments/assets/76356a3c-4911-4531-86ec-905d1702bfc0)

**Comment by LoganBooker** [comment]
@betterftr I wasn't able to replicate the super dampened LR growth you're seeing -- until I included TE2 and TE3. This absolutely kills the LR growth.

Can you try including just TE1 and see if that makes a difference?

**Comment by betterftr** [comment]
Interesting even tho TE1, TE2 and TE3 is included, their weights should not be trained unless turning on their train switches in OT, my tests:
Orange: split_groups=True, d0=1e-06
Blue: split_groups=True, d0=1e-05
Deep orange: split_groups=True, d0=1e-05, only TE1 included
![image](https://github.com/user-attachments/assets/24662f10-716f-439b-ad57-6b62f1f74bb8)

**Comment by LoganBooker** [comment]
Hmm, interesting. Thanks for trying those suggestions. At this stage I might be limited in terms of diagnosing specifically what's happening with Large that isn't with Medium. All I can think of is maybe try setting some things to defaults (like timestep, loss weighting, etc.) and making sure none of those are messing with the calculations. That said, I've tried various combinations of those on my setup, and the optimiser is able to start adjusting the LR within 5-10 steps.

Will keep trying what I can!

**Comment by LoganBooker** [comment]
@betterftr I've just pushed a hack that should help with the LR. It's not strictly correct, but it should help while I figure out the root cause.

**Comment by betterftr** [comment]
I am downloading sd3.5 medium rn so I can test with that too

**Comment by LoganBooker** [comment]
Yep, I think I know what's going on. If you try the latest commit it should do the trick (for now). I'm still working on the root cause.

**Comment by betterftr** [comment]
yep, it did the trick (3.5M) all TE included, bs4
![image](https://github.com/user-attachments/assets/59cf4bb2-71ff-451c-bd91-3f1ec7183893)

**Comment by LoganBooker** [comment]
Yep, like I said it's a bit of a hack. :P For now, see where it plateaus in the first 50-100 steps, and set `prodigy_steps` to that value to turn Prodigy off after that point.

**Comment by LoganBooker** [comment]
@betterftr So, the root cause is the accumulation of negative values in the numerator early in training. Because these values persist with a very small decay (0.995-0.999 in most Prodigy configurations), the learning rate basically never increases, or only does so very late in training.

The fix is to clamp numerator updates to 0 -- this has the same effect of stopping the numerator from growing when it shouldn't, while also stopping it from being sabotaged in early training.

The second part is that, even with this fix, the numerator grows very, very slowly. While you can try training with just this fix, I'd recommend bumping up `d_coef` to `10`. You should see the LR start to increase within ~20 steps.

**Comment by betterftr** [comment]
Morning, should I set the new parameter "abs_numerator" to something else other than the default False? My lr seems to be going up a bit high still, also with beta4 None, should I experience seeing my lr go down too?
![image](https://github.com/user-attachments/assets/21db967d-5030-4839-8fd8-33406398e1e1)

![image](https://github.com/user-attachments/assets/482c9391-812a-4187-88b6-bc3791a99e20)

**Comment by LoganBooker** [comment]
`abs_numerator` is the hack I added from before. I've implemented what I think is a proper fix (the clamping mentioned in my other comment), but I left it as a toggle in case anyone wanted to experiment with it. I'd suggest leaving it `False`, and using `d_coef` if the LR is slow to increase.

The reference Prodigy implementation prevents `d` (the LR) from ever going lower -- it'll always increase. Ideally it should only take a little bit of training time for it to shoot up to the optimal LR (or close to it). From my experience, this isn't a problem in regular models, but diffusion models have different loss/gradient behaviour, which is why we've always had to pair it with a cosine/linear scheduler to prevent the behaviour you're seeing. 

Fortunately, the schedule-free component means we don't need the LR to decrease -- it's just needs to be set at the best value from the outset or very early and stay there. To do that with Prodigy, we need to turn it off at an appropriate time, which is where the `prodigy_steps` setting comes in.

So, for now I'd recommend:
`abs_numerator=False,beta4=0,d_coef=10,prodigy_steps=50` -- though `prodigy_steps` might be best set to the same number of steps as a single epoch.

**Comment by betterftr** [comment]
okay. so basically if I dont use prodigy_steps my lr keeps increasing indefinitely for now?

**Comment by LoganBooker** [comment]
Yep. FYI, schedule-free can require an LR 10x that of regular AdamW, so the high LRs you're seeing might be normal.

**Comment by betterftr** [comment]
> Yep. FYI, schedule-free can require an LR 10x that of regular AdamW, so the high LRs you're seeing might be normal.

Nah I sampled they are fried

**Comment by LoganBooker** [comment]
In that case, probably not normal!

This is the weight norm of an SDXL LoRA trained using schedule-free. As you can see, it naturally starts to peter out, hence why we don't need an external schduler.

![image](https://github.com/user-attachments/assets/13511cf5-bab9-420b-9aa4-3bd1eb1f9cf9)

**Comment by betterftr** [comment]
whelp, prodigy_steps 100 was too much will try 50
![image](https://github.com/user-attachments/assets/72e70cbf-049c-4323-be85-c95f72eac5a7)
![image](https://github.com/user-attachments/assets/89a25832-889d-4fd0-9e38-e35a9a539fb7)

**Comment by LoganBooker** [comment]
It actually looks like you're getting a good impulse at the start from that graph, so you could try using a lower `d_coef`. A value of `2` might be enough.

**Comment by betterftr** [comment]
> lower `d_coef`

that was with 1.0 default

**Comment by LoganBooker** [comment]
Wow, fair enough! I use a value of 'prodigy_steps=20` for my training -- you could probably get away with 10 here.

**Comment by betterftr** [comment]
well rip :D something else is wrong rather

![image](https://github.com/user-attachments/assets/05e5ef16-3fa6-41f5-92af-871c187a9f68)
![2024-11-16_11-08-22-training-sample-150-5-0](https://github.com/user-attachments/assets/b5fac586-a62d-48df-976a-3155344e76fb)
![image](https://github.com/user-attachments/assets/3b2699ad-8b82-47b6-93ef-75020694990c)

**Comment by LoganBooker** [comment]
@betterftr Yep, I agree. I think there's something else going on -- I'm not seeing anything like this myself.

**Comment by betterftr** [comment]
![image](https://github.com/user-attachments/assets/9829f642-206b-49d7-8cdd-6a16f465fc6f)
![2024-11-16_11-22-19-training-sample-150-5-0](https://github.com/user-attachments/assets/fbfa01b7-0f6e-4215-a978-91324083af8f)

prodigy_steps=20 ended up with 1.9e-05 and it seems the samples are still fried
will try your suggestion now and get a baseline

**Comment by betterftr** [comment]
While trying to get baseline with Adafactor, I found something interesting in OT. There seems to be a problem with cranking up CPU offloading to 1.0 + Adafactor that might cause all this
test1, samples not fried, Adafactor 1.9e-05 lr, bs4
![image](https://github.com/user-attachments/assets/ae3c0808-811a-46c4-94f7-a9cce4055650)

![image](https://github.com/user-attachments/assets/540b9c61-d47f-472c-b0a5-8ef01991483f)
![image](https://github.com/user-attachments/assets/a282b29f-0cd5-4b88-a2c4-29964273ed11)

test2, samples fried, Adafactor 1.9e-05 lr, bs4
![image](https://github.com/user-attachments/assets/96bd1fc5-e519-4de6-ae4b-2959dcba3767)
![image](https://github.com/user-attachments/assets/3cca217d-dc63-42ee-bcee-a950ccb7a842)
![image](https://github.com/user-attachments/assets/e85a9a6a-0f82-40e4-9eae-607104a0d9c0)

and test3 samples not fried, Adafactor 1.9e-05 lr, bs4 layer offload 0.7

**Comment by betterftr** [comment]
ah yea its a different story now.

Orange: abs_numerator=False
Blue: abs_numerator=True
![image](https://github.com/user-attachments/assets/4ce50050-ec7a-4c2e-8653-418e3e0c25f6)
And samples are fine when I set prodigy_steps=20

**Comment by LoganBooker** [comment]
@betterftr Glad you were able to get to the bottom of it. I've been doing some more investigating myself, and it seems that while Adam-atan2 works completely fine with SDXL (well, LoRAs at least), it seems to interfere with Prodigy quite a bit in newer models / finetuning.

In the latest commit, I've made the default to be regular epsilon, which seems to have much better LR adaptation performance.

I also removed the `abs_numerator` parameter, as it is very incorrect in terms of how Prodigy is meant to operate. It also seems unneeded now I believe.

**Comment by betterftr** [comment]
Nice :) so the only thing there is to figure out is how to get rid of prodigy steps and make it behave like og prodigy, finding its plateau?

**Comment by LoganBooker** [comment]
@betterftr Prodigy's natural behaviour is to increase the LR over time, it's just been hidden for the most part as it's usually paired with a decaying cosine/linear scheduler.

That said, the most recent commits should have much better plateauing behaviour. It doesn't run as "hot" as previous commits, but you should see the LR increasing within ~50 steps (at batch counts of 1-2 on Medium at least).
